kvm-xfstests environment improvements

dmonakhov opened the pull request:
Hi,

Please review my patchset. It's primary idea is to improve configuration flexibility.
Note: last cleanup patch allow to reduce kvm-xfstests docker environment from 300Mb to 174M so docker runtime environment becomes closer to optimum size.

tytso commented:
None of these commits have a signed-off-by.  Can you rebase with signed-off-by, or send a note here giving me permission to add your signed-off-by?

dmonakhov commented:
Ok. updated against master and add sign-offs

tytso commented:
Regarding the last commit (the one that changes kvm-xfstests/Dockerfile to use Alpine).   

Does the usage command really work?  Last I tried running qemu in Docker, it requires --privileged, and it was a gaping security hole.  So it never really seemed to be viable for anything more than as a party trick, to be honest.

As opposed the Dockerfile in the top-level, which actually is useful for doing test builds, and also as a way of testing the file system in whatever Docker environment provides, and which does not require any special privileges.

dmonakhov commented:
Hm... In fact all it needs is to grant  access to /dev/kvm so 'docker --device /dev/kvm:/dev/kvm' should be enough 

That's depends of your environment. If you try to run in on public cloud then yes, this is security hole.
But it you run in on your cluster which you (or your team) own  then this is very usefull. For example I've join old crappy boxes to one kubernetes cluster which aggregates 300 vcpu and 1Tbmem . I split xfstests to ~100 independent jobs and run it in paralles as kubernetes-jobs. This allow me to complete 'smoke' scenario ~3min.  IMHO this is fantastic achievement which may be usefull for others.

dmonakhov commented:
Yeah. It seems that running qemu-kvm inside container is now safe in GCE :) https://cloudplatform.googleblog.com/2017/09/introducing-nested-virtualization-for.html
https://cloud.google.com/compute/docs/instances/enable-nested-virtualization-vm-instances

tytso commented:
Sure, but nested virtualization doesn't buy us much over simply running xfstests in GCE natively.  That's what gce-xfstests is all about.    In fact, thanks to some great work by a summer intern, we now have now have test manager so that we can break up the work to muliple VM's, monitor the VM's to see if they wedge, capture the serial console output, etc.   At the moment we are only breaking up the work so that it's one file system config per VM, but it allows us to collapse 23 hours (from running "gce-xfstests -c ext4/all -g auto") down to 3 hours.    We have thought about breaking up the work in a more finer-grained way, but there are some tradeoffs.  In particular, if we shard in a too-fine-grained way, then lockdep bugs and memory bugs are much less likely to be detected.    That might be OK for a smoke test, but for a full regression test, I'd want to stick with one VM per file system config.

Back to the original question.  If the only thing needed is "kvm-xfstests -p /dev/kmem:/dev/kmem", shouldn't that be mentioned in the usage message listed in the Dockerfile?

amir73il reviewed (commented):
typo in subject s/eraly/early

amir73il commented on kvm-xfstests/config.kvm:
```diff
@@ -40,3 +41,11 @@ MONITOR="-monitor telnet:localhost:7498,server,nowait"
+# The issue has not been observed with QEMU 2.9.0.
+# Native AIO is probably better for our goal which is test proofing
+# production systems.
+AIO=",aio=native"
```
minor suggestion, maybe keep this up near CACHE and the disks configuration?

dmonakhov replied:
Reasonable. Thank you.

dmonakhov commented:
Ok. I've got what you are talking about. Yes, usage definitely need to be updated. See updated pathes

Commit:
gen-image: fail early on errors

Image creation is rare and important process, any error during generation
potentially result in bad image which result in unpredictable behaviour on
test execution. Let's create image creation supersafe.

Signed-off-by: Dmitry Monakhov <dmonakhov@openvz.org>

Commit:
kvm-xfstests: add qemu vdisk config option

In most cases we do not care about true integrity, but we do care about
test time, cache=unsafe allow to speedup test execution significantly.

NONE: Origianal default behaviour(cache=none) preserved.

Signed-off-by: Dmitry Monakhov <dmonakhov@openvz.org>

Commit:
docker: update kvm-xfstests runtime docker image

- Use https://alpinelinux.org image as a base. This allow us reduce image size
  almost x2 times.
- Update obsolete comment. This dockerfile create runtime environment.

Signed-off-by: Dmitry Monakhov <dmonakhov@openvz.org>

Commit:
kvm-xfstests: add result archive generation

It would be nice have an ability to fetch detailed logs from result vdisk image
but in order to do that we need to mount it, which annoying. Even more
this not works for environments where we have not mount permissions or
access to /dev/loop.  Let's archive tests results to single file and fetch
it from disk image via debugfs.
Example:
#./kvm-xfstests.sh --archive -c 4k generic/001
...
result  in ./logs/resutls-201704081703.tar.xz
logfile in ./logs/log.201704081703

Signed-off-by: Dmitry Monakhov <dmonakhov@openvz.org>

Commit:
{gce,kvm}-xfstess: Pass extra options for xfstests script

./check script has many options, let's allow user to use it.

Example: Not run tests, just list tests from 'smoke' category
# kvm-xfstests -O -n smoke

Signed-off-by: Dmitry Monakhov <dmonakhov@openvz.org>

Commit:
{kvm,gce}-xfstests: make virtfs-9p path configurable

Signed-off-by: Dmitry Monakhov <dmonakhov@openvz.org>

Commit:
ci: add basic travic-ci script

travis-ci.org is one of free and popular ci services with transparent
integration for github. In order to enable automatic build-ci on
new commit or poll-request just toggle it at:
https://github.com/USER/xfstests-bld/settings/installations

dmonakhov commented:
I've add one more patch which adds travis-ci script. I've copy-pasted it from fio.git . It is very trivial but still usefull. Here is example of build-log https://travis-ci.org/dmonakhov/xfstests-bld/builds/281336943

tytso commented:
OK, I've integrated all of your changes into my tree.

Note that Travis CI is requires that you give them read:org privileges, which means they get to see basic information about all of your organization's private repositories.   Your company have rules which absolutely prohibit leakage of the companies private informaiton to a third party, such as Travis CI.   At least for me, doing continuous integration *after* I've made the git repo published for all of the world to see doesn't actually help all that much.   So I just tend to do a full build *plus* trying to run a few smoke tests before I push changes out publically --- and so Travis doesn't actually buy me much, personally.
